### DISCUSSION TITLE
Json-server sendo usado como API em uma aplicação angular

### DISCUSSION THREAD

**Discussion by joao-lim4** [frontendbr/forum#1809] [Discussão]
Eu tenho uma aplicação funcionando usando php e mysql, porém queria migrar ela para o angular refazer todo o design dela e tals porém nunca fiz um crud no angular usando mysql e foi ai que pensei em usar o json-server como api pra ele pois eu consigo manipular tudo que eu quiser, porem eu não sei se é viável usar o json-server em uma aplicação. Tem algumas ideias de banco de dados que eu possa usar com o angular ?? Aceito dicas e conselhos kkkk

**Comment by lddv** [comment]
Olá, @joao-lim4 . O Angular, até onde sei, não tem nenhuma dependência com qualquer banco de dados. Ele é um framework front-end, portanto você é livre pra usar o que quiser no back-end e no banco de dados.

Sobre o [json-server](https://www.npmjs.com/package/json-server), usá-lo significa introduzir uma camada Node.js pra fazer a API e ler os dados de um json fictício. Se quiser testar os métodos da API sem preocupar em apagar dados reais do seu banco, acho válida essa opção.

No entanto, para um resultado final mais próximo ao que você já tem hoje, reaproveitar o MySQL no seu caso de parece a melhor opção. Se quiser brincar um pouco mais e sair do paradigma relacional, fique à vontade para se aventurar com o MongoDB.

Espero ter ajudado. Um abraço.

**Comment by wilsonneto-swe** [comment]
@joao-lim4 se você ver as primeiras linhas do repo do json-server, os desenvolvedores escreveram:
"Created with <3 for front-end developers who need a quick back-end for prototyping and mocking."

Bom... **prototyping and mocking.**, eu uso demais ele apenas para isso, prototipar, fazer meu backend com dados fake e simulando uma api real, depois faço minha api com um banco relacional e na maior parte das vezes com express.js, nest.js ou .net core. Nunca usei e acho melhor nunca usar json-server em prod.

Abraços!

**Comment by ninetails** [comment]
O ideal seria usar apenas em desenvolvimento mesmo, como apontado pelo @wilsonneto-dev .

Mas creio que vc consiga encontrar frameworks que trabalhem somente com APIs na parte do backend. A evolução da tua arquitetura deveria ser em desacoplar backend do frontend. E a parte do backend, seria encontrar um outro framework que faria tal interface pra ti... ou desacoplar a certo ponto que poderia ser possível até usar algo pronto como o Firebase.

Sei que no Express vc pode retornar json e ele retorna como texto (apesar de achar que provavelmente precisa explicitar o _mime-type_ pra `Application/json` no header,  se n me engano), e daí trabalhar cada caso do CRUD [de acordo com a rota](https://expressjs.com/en/starter/basic-routing.html).

Mas como disse, tendo um contrato, vc pode utilizar outras stacks pro seu backend. Tomando REST como exemplo, aí vc pode mudar qualquer intermediário entre sua aplicação frontend e o banco de dados... fazer o backend em Laravel com PHP, Express no NodeJS, Django no Python... Ou mesmo online: [Firebase](https://firebase.google.com/docs/reference/rest/database), [Amazon REST API](https://docs.aws.amazon.com/apigateway/api-reference/) ou qualquer outra solução na nuvem, desde que mantendo-se o contrato.

Mas como a galera disse, o Angular é somente um framework pro frontend que serve de interface web para sua aplicação. O ideal é comunicar com o backend utilizando uma API feita de modo apartado... que pessoalmente creio que seja a tendência atual.

Fazendo desacoplado e não mudando contratos vc tem a chance de qualquer coisa no backend e poder mudar/atualizar a qualquer momento. O mesmo pro frontend. E aí você pode variar a tecnologia/stack/framework de um sem influir no outro, pq no final seria só mudar o endpoint.

**Comment by coffeeniac** [comment]
@joao-lim4 a ideia do json-server é utilizar em dev mesmo para testes rápidos ou para criação de componentes com chamadas 'fake' e depois vincular a um end-point válido apontando para um API mesmo.

Caso queira montar uma API com pouca curva de aprendizado, já que você está trabalhando com Angular, aconselho usar NodeJS, como o @ninetails apontou, que tem uma sintaxe fácil para montar uma API e lá você faria a conexão com um banco de dados e etc.
